add test to demonstrate kubernetes-sigs/kustomize#1835

muenchdo opened the pull request:
This adds a new test to demonstrate kubernetes-sigs/kustomize#1835 as described in the guide on how to [file bugs](https://github.com/kubernetes-sigs/kustomize/blob/master/docs/bugs.md).

muenchdo commented:
/assign @droot

monopole commented on api/krusty/mergeenvfrom_test.go:
```diff
@@ -0,0 +1,67 @@
+  name: nginx
+spec:
+  template:
+    spec:
```
in lists, we line up the hyphens with the list name, e.g.
```
      containers:
      - name: nginx
        image: image1
        envFrom:
        - configMapRef:
            name: some-config
        - configMapRef:
            name: more-config
```

monopole commented on api/krusty/mergeenvfrom_test.go:
```diff
@@ -0,0 +1,67 @@
+// When patching, `envFrom` should merge the list instead of replacing it.
+func TestMergeEnvFrom(t *testing.T) {
+	th := kusttest_test.MakeHarness(t)
+	th.WriteF("/app/base/deployment.yaml", `
```
This could be just `"app/deployment.yaml"`

monopole commented on api/krusty/mergeenvfrom_test.go:
```diff
@@ -0,0 +1,67 @@
+            - configMapRef:
+                name: more-config
+`)
+	th.WriteK("/app/base", `
```
just `"app",`

monopole commented on api/krusty/mergeenvfrom_test.go:
```diff
@@ -0,0 +1,67 @@
+              - configMapRef:
+                  name: another-config
+`)
+	m := th.Run("/app/base", th.MakeDefaultOptions())
```
likewise just `"app",`

There's no overlay, so no need to add a directory level.

monopole commented on api/krusty/mergeenvfrom_test.go:
```diff
@@ -0,0 +1,67 @@
+        image: image1
+        name: nginx
+`)
+}
```
Please add the following to the test, see if if does the job you want
```
func TestMergeEnvFromViaJsonInline(t *testing.T) {
	th := kusttest_test.MakeHarness(t)
	th.WriteF("app/deployment.yaml", `
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx
spec:
  template:
    spec:
      containers:
      - name: nginx
        image: image1
        envFrom:
        - configMapRef:
            name: some-config
        - configMapRef:
            name: more-config
`)
	th.WriteK("app", `
resources:
- deployment.yaml
patches:
- target:
    kind: Deployment
    name: nginx
  patch: |-
    - op: add
      path: /spec/template/spec/containers/0/envFrom/-
      value:
        configMapRef:
          name: another-config
`)
	m := th.Run("app", th.MakeDefaultOptions())
	th.AssertActualEqualsExpected(m, `
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx
spec:
  template:
    spec:
      containers:
      - envFrom:
        - configMapRef:
            name: some-config
        - configMapRef:
            name: more-config
        - configMapRef:
            name: another-config
        image: image1
        name: nginx
`)
}

```

muenchdo replied:
That does the trick 🙂

monopole commented:
/ok-to-test

muenchdo commented:
Hi @monopole, thanks for that super-fast review! I addressed your comments and also factored the generation of the deployment manifest into a separate function. Your suggestion of using a JSON6902 patch worked 🙂
